src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +616,289 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
 }
 
 
+/*
+ * RecursivelyPlanRecurringTupleOuterJoins descends into a join tree and
+ * recursively plans all non-recurring (i.e., distributed) rels that that
+ * participate in an outer join expression together with a recurring rel,
+ * such as <distributed> in "<recurring> LEFT JOIN <distributed>", i.e.,
+ * where the recurring rel causes returning recurring tuples from the worker
+ * nodes.
+ */
+static void
+RecursivelyPlanRecurringTupleOuterJoins(Node *node, Query *query,
+										RecursivePlanningContext *recursivePlanningContext)
+{
+	if (node == NULL)
+	{
+		return;
+	}
+	else if (IsA(node, RangeTblRef))
+	{
+		return;
+	}
+	else if (IsA(node, FromExpr))
+	{
+		FromExpr *fromExpr = (FromExpr *) node;
+		ListCell *fromExprCell;
+
+		/* search for join trees in each FROM element */
+		foreach(fromExprCell, fromExpr->fromlist)
+		{
+			Node *fromElement = (Node *) lfirst(fromExprCell);
+
+			RecursivelyPlanRecurringTupleOuterJoins(fromElement, query,
+													recursivePlanningContext);
+		}
+
+		return;
+	}
+	else if (IsA(node, JoinExpr))
+	{
+		JoinExpr *joinExpr = (JoinExpr *) node;
+
+		Node *leftNode = joinExpr->larg;
+		Node *rightNode = joinExpr->rarg;
+
+		/* there may be recursively plannable outer joins deeper in the join tree */
+		RecursivelyPlanRecurringTupleOuterJoins(joinExpr->larg, query,
```

Comment:
to trigger this, lets make sure to add tests like: 

```SQL
SELECT count(*) FROM test_r_1 LEFT JOIN (  (test_r_2  JOIN test_1 USING (key))  JOIN  (test_2  JOIN test_3 USING (key)) USING(key))  USING (key);
```

please write even deeper join trees. The above only works when we merge @aykut-bozkurt's PR.

But, we should also have with a top level subquery:
```SQL
 SELECT count(*) FROM test_r_1 LEFT  JOIN ( SELECT * FROM  (SELECT * FROM test_r_2 JOIN test_1 USING (key)) as foo  JOIN  (SELECT * FROM test_r_3 LEFT JOIN test_3 USING (key)) as bar USING(key)) as zoo  USING (key);
```